protocol/binary.go

```diff
@@ -61,3 +82,108 @@ func (binaryProtocol) DecodeEnveloped(r io.ReaderAt) (wire.Envelope, error) {
 	e, err := reader.ReadEnveloped()
 	return e, err
 }
+
+// DecodeRequest specializes Decode and replaces DecodeEnveloped for the
+// specific purpose of decoding request structs that may or may not have an
+// envelope.
+// This allows a Thrift request handler to transparently accept requests
+// regardless of whether the caller submits an envelope.
+func (b binaryProtocol) DecodeRequest(r io.ReaderAt) (wire.Value, EnvelopeSpecificResponder, error) {
+	// Strip request envelopes if present.
+	// 1. A message of length 1 containing only 0x00 can only be a bare,
+	// un-enveloped empty struct.
+	// 2. A message of length >1 starting with 0x00 can only be enveloped
+	// without a version (non-strict)
+	// 3. A message of length >1 with MSB 0x80 set can only be enveloped with a
+	// version (strict)
+	// 4. A message of length >1 starting with 0x00-0x0f can only be the
+	// beginning of an un-enveloped field.
+	var buf [2]byte
+	// If we fail to read two bytes, the only possible valid value is the empty struct.
+	if _, err := r.ReadAt(buf[0:2], 0); err != nil {
+		val, err := b.Decode(r, wire.TStruct)
+		if err != nil {
+			return wire.Value{}, noEnvelopeResponder, err
+		}
+		return val, noEnvelopeResponder, nil
+	}
+
+	// 0x00 is a valid preamble for a non-strict enveloped request.
+	if buf[0] == 0x00 {
+		e, err := b.DecodeEnveloped(r)
+		if err != nil {
+			return wire.Value{}, noEnvelopeResponder, err
+		}
+		return e.Value, &envelopeV0Responder{
+			Name:  e.Name,
+			SeqID: e.SeqID,
+		}, nil
+	}
+
+	// 0x80 is a valid preamble for a strict enveloped request.
+	if buf[0]&0x80 > 0 {
+		e, err := b.DecodeEnveloped(r)
+		if err != nil {
+			return wire.Value{}, noEnvelopeResponder, err
+		}
+		return e.Value, &envelopeV1Responder{
+			Name:  e.Name,
+			SeqID: e.SeqID,
+		}, nil
+	}
+
+	// All other patterns are either bare structs or invalid.
+	val, err := b.Decode(r, wire.TStruct)
+	if err != nil {
+		return wire.Value{}, noEnvelopeResponder, err
+	}
+	return val, noEnvelopeResponder, nil
+}
+
+// noEnvelopeResponder responds to a request without an envelope.
+
+type _noEnvelopeResponder struct{}
+
+func (_noEnvelopeResponder) EncodeResponse(v wire.Value, t wire.EnvelopeType, w io.Writer) error {
```

Comment:
should we go the other way (underscore prefix for the variable, and not the type, since that also matches our internal style guide)